### PR TITLE
Curiosity Driven Exploration & Pyramids Environments

### PR DESCRIPTION
**Pull request by awjuliani**
* Adds implementation of *Curiosity-driven Exploration by Self-supervised Prediction* (https://arxiv.org/abs/1705.05363) to PPO trainer.
* To enable, set `use_curiosity` flag to `true` in hyperparameter file.
* Includes refactor of `unitytrainers` model code to accommodate new feature.
* Adds new `Pyramids` environment (w/ documentation). Environment contains sparse reward, and can only be solved using PPO+Curiosity.

In particular looking for feedback on python-side refactoring, as there may be nicer way of doing things.

### CODE REVIEW

**Comment by mmattar** [comment]
@awjuliani How do the results change by enabling curiosity for our current environments?

File: python/unitytrainers/models.py [hunk 1]
```diff
@@ -44,109 +56,137 @@ def create_visual_input(o_size_h, o_size_w, bw, name):
         visual_in = tf.placeholder(shape=[None, o_size_h, o_size_w, c_channels], dtype=tf.float32, name=name)
         return visual_in
 
-    def create_vector_input(self, s_size):
+    def create_vector_input(self):
```
Comment (vincentpierre) [inline python/unitytrainers/models.py]: Should `create_vector_input` take in a name as argument in the same way `create_visual_input` does ?

File: python/unitytrainers/models.py [hunk 2]
```diff
@@ -159,7 +199,8 @@ def create_new_obs(self, num_streams, h_size, num_layers):
             final_hiddens.append(final_hidden)
         return final_hiddens
 
-    def create_recurrent_encoder(self, input_state, memory_in, name='lstm'):
+    @staticmethod
+    def create_recurrent_encoder(input_state, memory_in, sequence_length, name='lstm'):
         """
```
Comment (vincentpierre) [inline python/unitytrainers/models.py]: For uniformity, maybe the name in all those methods should be an optional argument (same with reuse) ?

File: python/unitytrainers/ppo/models.py [hunk 3]
```diff
@@ -44,6 +49,92 @@ def create_reward_encoder():
         update_reward = tf.assign(last_reward, new_reward)
         return last_reward, new_reward, update_reward
 
+    def create_curiosity_encoders(self):
```
Comment (vincentpierre) [inline python/unitytrainers/ppo/models.py]: Maybe for latter : Each encoder could be in its own script ?

File: python/unitytrainers/ppo/models.py [hunk 4]
```diff
@@ -44,6 +49,92 @@ def create_reward_encoder():
...
+                next_visual_encoders.append(encoded_next_visual)
+
+            hidden_visual = tf.concat(visual_encoders, axis=1)
+            hidden_next_visual = tf.concat(next_visual_encoders, axis=1)
+            inverse_input_list.append(hidden_visual)
+            inverse_input_list.append(hidden_next_visual)
+            encoded_state_list.append(hidden_visual)
+            encoded_next_state_list.append(hidden_next_visual)
+
+        if self.o_size > 0:
```
Comment (vincentpierre) [inline python/unitytrainers/ppo/models.py]: I think we have a better way to check for continuous vector observation than than. It might cause problems in scenarios where the vector observation is a continuous single float.

File: python/unitytrainers/ppo/models.py [hunk 5]
```diff
@@ -44,6 +49,92 @@ def create_reward_encoder():
...
         return last_reward, new_reward, update_reward
 
+    def create_curiosity_encoders(self):
+        """
+        Creates state encoders for current and future observations.
+        Used for implementation of ﻿Curiosity-driven Exploration by Self-supervised Prediction
+        See https://arxiv.org/abs/1705.05363 for more details.
+        :return: current and future state encoder tensors.
+        """
+        inverse_input_list = []
```
Comment (vincentpierre) [inline python/unitytrainers/ppo/models.py]: I cannot see where this is used. Where is this list consumed ?

File: python/unitytrainers/ppo/models.py [hunk 6]
```diff
@@ -44,6 +49,92 @@ def create_reward_encoder():
...
+            encoded_state_list.append(encoded_vector_obs)
+            encoded_next_state_list.append(encoded_next_vector_obs)
+
+        encoded_state = tf.concat(encoded_state_list, axis=1)
+        encoded_next_state = tf.concat(encoded_next_state_list, axis=1)
+        return encoded_state, encoded_next_state
+
+    def create_inverse_model(self, encoded_state, encoded_next_state):
+        """
+        Creates inverse model TensorFlow ops for Curiosity module.
```
Comment (vincentpierre) [inline python/unitytrainers/ppo/models.py]: Maybe add something like : Predicts the action taken between state and next state.

File: python/unitytrainers/ppo/models.py [hunk 7]
```diff
@@ -44,6 +49,92 @@ def create_reward_encoder():
...
+        :param encoded_state: Tensor corresponding to encoded current state.
+        :param encoded_next_state: Tensor corresponding to encoded next state.
+        """
+        combined_input = tf.concat([encoded_state, encoded_next_state], axis=1)
+        if self.brain.vector_action_space_type == "continuous":
+            pred_action = tf.layers.dense(combined_input, self.a_size, activation=None)
+            squared_difference = tf.reduce_sum(tf.squared_difference(pred_action, self.selected_actions), axis=1)
```
Comment (vincentpierre) [inline python/unitytrainers/ppo/models.py]: Where is self.selected_actions created in continuous control. In discrete, it is in create_dc_actor_critic. I think self.selected_actions should be in create_dc_actor_critic then. Did you try curiosity with a continuous control environment to make sure self.selected_action is used ?

File: python/unitytrainers/ppo/trainer.py [hunk 8]
```diff
@@ -231,31 +260,41 @@ def add_experiences(self, curr_all_info: AllBrainInfo, next_all_info: AllBrainIn
...
+                        self.training_buffer[agent_id]['actions_pre'].append(actions_pre[idx])
                     a_dist = stored_take_action_outputs[self.model.all_probs]
                     value = stored_take_action_outputs[self.model.value]
                     self.training_buffer[agent_id]['actions'].append(actions[idx])
-                    if self.is_continuous_action:
-                        self.training_buffer[agent_id]['actions_pre'].append(actions_pre[idx])
                     self.training_buffer[agent_id]['prev_action'].append(stored_info.previous_vector_actions[idx])
                     self.training_buffer[agent_id]['masks'].append(1.0)
-                    self.training_buffer[agent_id]['rewards'].append(next_info.rewards[next_idx])
+                    if self.use_curiosity:
```
Comment (vincentpierre) [inline python/unitytrainers/ppo/trainer.py]: Would it make more sense to always have an intrinsic reward but only have it non-zero when curiosity is used ?

File: python/unitytrainers/ppo/trainer.py [hunk 9]
```diff
@@ -345,7 +390,7 @@ def update_model(self):
         """
         num_epoch = self.trainer_parameters['num_epoch']
         n_sequences = max(int(self.trainer_parameters['batch_size'] / self.sequence_length), 1)
-        total_v, total_p = [], []
+        total_v, total_p, total_f, total_i = [], [], [], []
```
Comment (vincentpierre) [inline python/unitytrainers/ppo/trainer.py]: Maybe it is time to find better names for these variables or is this outside of the scope of this PR ?

File: unity-environment/Assets/ML-Agents/Scripts/Agent.cs [hunk 10]
```diff
@@ -675,6 +675,11 @@ protected void AddVectorObs(Quaternion observation)
         info.vectorObservation.Add(observation.z);
         info.vectorObservation.Add(observation.w);
     }
+    
+    protected void AddVectorObs(bool observation)
```
Comment (vincentpierre) [inline unity-environment/Assets/ML-Agents/Scripts/Agent.cs]: Add a `<summary>`

File: unity-environment/Assets/ML-Agents/Examples/Pyramids/Scripts/PyramidSwitch.cs [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+    private bool state;
+    private GameObject area;
+    private PyramidArea areaComponent;
+
+    public bool GetState()
+    {
+        return state;
+    }
+
+    private void Start()
```
Comment (vincentpierre) [inline unity-environment/Assets/ML-Agents/Examples/Pyramids/Scripts/PyramidSwitch.cs]: InitializeAgent?

File: unity-environment/Assets/ML-Agents/Examples/Pyramids/Scripts/PyramidSwitch.cs [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+    {
+        area = gameObject.transform.parent.gameObject;
+        areaComponent = area.GetComponent<PyramidArea>();
+        Reset();
```
Comment (vincentpierre) [inline unity-environment/Assets/ML-Agents/Examples/Pyramids/Scripts/PyramidSwitch.cs]: Can we name Reset ResetSwitch ?

File: python/unitytrainers/ppo/models.py [hunk 13]
```diff
@@ -44,6 +49,92 @@ def create_reward_encoder():
...
+            inverse_input_list.append(encoded_vector_obs)
+            inverse_input_list.append(encoded_next_vector_obs)
+            encoded_state_list.append(encoded_vector_obs)
+            encoded_next_state_list.append(encoded_next_vector_obs)
+
+        encoded_state = tf.concat(encoded_state_list, axis=1)
+        encoded_next_state = tf.concat(encoded_next_state_list, axis=1)
+        return encoded_state, encoded_next_state
+
+    def create_inverse_model(self, encoded_state, encoded_next_state):
```
[the remark this answers is not in the document]
Comment (awjuliani) [reply python/unitytrainers/ppo/models.py]: Good point.

File: python/unitytrainers/ppo/models.py (hunk 3 above)
Comment (awjuliani) [reply python/unitytrainers/ppo/models.py]: Can you elaborate? Are you thinking of a `curiosity.py`?

File: python/unitytrainers/ppo/models.py (hunk 4 above)
Comment (awjuliani) [reply python/unitytrainers/ppo/models.py]: What would you suggest?

File: python/unitytrainers/ppo/trainer.py (hunk 8 above)
Comment (awjuliani) [reply python/unitytrainers/ppo/trainer.py]: If we did that we'd need some code to generate zeroes and keep track of them, which doesn't seem to be as elegant.

File: python/unitytrainers/ppo/trainer.py (hunk 9 above)
Comment (awjuliani) [reply python/unitytrainers/ppo/trainer.py]: I will make them less terrible here. It is justified since we are adding new ones.

File: unity-environment/Assets/ML-Agents/Examples/Pyramids/Scripts/PyramidSwitch.cs (hunk 12 above)
Comment (awjuliani) [reply unity-environment/Assets/ML-Agents/Examples/Pyramids/Scripts/PyramidSwitch.cs]: Yes. We. Can.

File: python/unitytrainers/ppo/models.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (awjuliani) [reply python/unitytrainers/ppo/models.py]: Actually, this may require a larger refactoring. I think it would be best to wait as part of a our larger changes we will be making to trainers in the future.

File: python/unitytrainers/ppo/models.py [hunk 14]
```diff
@@ -35,6 +36,8 @@ def __init__(self, brain, lr=1e-4, h_size=128, epsilon=0.2, beta=1e-3, max_step=
         else:
             self.create_dc_actor_critic(h_size, num_layers)
         if self.use_curiosity:
+            self.encoding_size = encoding_size
```
Comment (vincentpierre) [inline python/unitytrainers/ppo/models.py]: encoding_size could have a more explicit name.

File: python/unitytrainers/ppo/models.py [hunk 15]
```diff
@@ -35,6 +36,8 @@ def __init__(self, brain, lr=1e-4, h_size=128, epsilon=0.2, beta=1e-3, max_step=
...
+            self.curiosity_strength = curiosity_strenght
```
Comment (vincentpierre) [inline python/unitytrainers/ppo/models.py]: typo

File: python/trainer_config.yaml [hunk 16]
```diff
@@ -17,13 +17,22 @@ default:
     sequence_length: 64
     summary_freq: 1000
     use_recurrent: false
+    use_curiosity: false
```
[the remark this answers is not in the document]
Comment (awjuliani) [reply python/trainer_config.yaml]: There is no model learned when `use_curiosity` is false. Are you suggesting we don't learn the model when strength == 0?

[the remark this answers is not in the document]
Comment (vincentpierre) [reply python/trainer_config.yaml]: Yes, and 0 is the default.

File: python/unitytrainers/ppo/models.py (hunk 14 above)
Comment (vincentpierre) [reply python/unitytrainers/ppo/models.py]: `curiosity_encoding_size` or `forward_model_h_size` or`curiosity_h_size`

File: python/unitytrainers/ppo/trainer.py [hunk 17]
```diff
@@ -43,6 +43,7 @@ def __init__(self, sess, env, brain_name, trainer_parameters, training, seed):
         self.use_recurrent = trainer_parameters["use_recurrent"]
         self.use_curiosity = bool(trainer_parameters['use_curiosity'])
         self.sequence_length = 1
+        self.has_updated = False
```
Comment (vincentpierre) [inline python/unitytrainers/ppo/trainer.py]: Can you put a comment describing what it does ?
I think it could also benefit a better name. From the code, I think it is something like `had_first_update`

File: unity-environment/ProjectSettings/TagManager.asset [hunk 18]
```diff
@@ -25,6 +25,7 @@ TagManager:
   - switchOff
   - pyramid
   - switchOn
+  - stone
```
Comment (vincentpierre) [inline unity-environment/ProjectSettings/TagManager.asset]: So many tags... Can we have themn more generic like `object0`, `object1`, `enemy0`...

File: python/unitytrainers/ppo/models.py [hunk 19]
```diff
@@ -8,7 +8,8 @@
 
 class PPOModel(LearningModel):
     def __init__(self, brain, lr=1e-4, h_size=128, epsilon=0.2, beta=1e-3, max_step=5e6,
-                 normalize=False, use_recurrent=False, num_layers=2, m_size=None):
+                 normalize=False, use_recurrent=False, num_layers=2, m_size=None, use_curiosity=False,
+                 curiosity_strength=0.01, encoding_size=128):
```
Comment (vincentpierre) [inline python/unitytrainers/ppo/models.py]: Please rename encoding_size to something containing curiosity
